src/MainWindow.vala

```diff
@@ -520,6 +523,33 @@ public class PantheonCalculator.MainWindow : Gtk.ApplicationWindow {
         }
     }
 
+    private void paste () {
+        get_clipboard_text.begin ((obj, res) => {
+            var text = get_clipboard_text.end (res);
+            if (text == null) {
+                return;
+            }
+
+            int start, end;
+            entry.get_selection_bounds (out start, out end);
+
+            var before = entry.text.slice (0, start);
+            var after = entry.text.slice (end, entry.text_length);
+            
+            entry.text = before + text + after;
+            entry.set_position (before.length + text.length);
```

Comment:
set_position is in char count which is different from byte length :slightly_smiling_face: 
```suggestion
            entry.set_position (before.char_count () + text.char_count ());
```